### PR TITLE
feat(masthead): masthead L1 web components

### PR DESCRIPTION
**Pull request by RaphaelAmadeu**
### Related Ticket(s)

#3806 

### Description

Initial version of Masthead L1 made based on the React version.

Some points that still need work are adding the items in the left menu and regarding the dropdown, as the react version uses the megamenu panel, that is in pr state for web components.

### CODE REVIEW

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 1]
```diff
@@ -126,6 +127,8 @@ class DDSMastheadComposite extends LitElement {
         });
   }
 
+  private
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead-composite.ts:130]: I guess you meant to remove this line?

```suggestion
```

**Review by asudoh — COMMENTED** [review]
Thank you for posting your in-progress work @RaphaelAmadeu! From my glance at the code, it appears that `<dds-masthead-l1>` is a similar component to `<dds-masthead-composite>` (rather than `<dds-masthead>`) with one difference, which is, rendering nav items and menu items in shadow DOM. For components like `<dds-masthead-composite>` which essentially are "render markup from data", there are two rules in the codebase:

* Render the content to light DOM instead of to shadow DOM (that said, we limit usage of shadow DOM to leaf components where encapsulation of shadow DOM works the most). By doing so:
  * Inspecting the light DOM tree tells you the semantic UI structure.
  * Inter-component communication, for example, one for hamburger menu button in top nav and the left nav, will be easier.
* Add `-composite` suffix to better tell that the component is for "render markup from data".

Hope it makes sense - Please ping me for any questions/concerns. Thanks!

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 2]
```diff
@@ -297,12 +389,14 @@ class DDSMastheadComposite extends LitElement {
...
             `}
         ${this._renderNavItems({ target: NAV_ITEMS_RENDER_TARGET.LEFT_NAV })}
+        ${ this.l1Data ? this._renderL1Items({ target: NAV_ITEMS_RENDER_TARGET.LEFT_NAV }) : undefined }
       </dds-left-nav>
       <dds-masthead aria-label="${ifNonNull(mastheadAssistiveText)}">
         <dds-masthead-menu-button
           button-label-active="${ifNonNull(menuButtonAssistiveTextActive)}"
           button-label-inactive="${ifNonNull(menuButtonAssistiveTextInactive)}"
         >
+        <slot></slot>
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead-composite.ts:399]: Contents in `<dds-masthead-composite>` is rendered to light DOM, and thus `<slot>` won't work. What did you want to achieve by `<slot>`?

**Review by asudoh — COMMENTED** [review]
Good progress @RaphaelAmadeu!

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 3]
```diff
@@ -68,6 +70,83 @@ enum NAV_ITEMS_RENDER_TARGET {
...
+      <dds-masthead-l1 slot="masthead-l1">
+        ${!title
+          ? undefined
+          : html`
+              <dds-masthead-l1-name title="${title}" url="${url}"> </dds-masthead-l1-name>
+            `}
+        ${this._renderL1Items({ target: NAV_ITEMS_RENDER_TARGET.TOP_NAV })}
+      </dds-masthead-l1>
+    `;
+  }
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead-composite.ts:148]: Given `l1Data` property seems optional, would you want to do:

```javascript
private _renderL1() {
  if (!this.l1Data) {
     return undefined;
  }
  ...
}
```

? The TypeScript error in the CI seems to be catching such condition. Similar thing seems to apply to `_renderL1Items()`.

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 4]
```diff
@@ -360,6 +439,12 @@ class DDSMastheadComposite extends LitElement {
   @property({ attribute: false })
   navLinks?: MastheadLink[];
 
+  /**
+   * Data for l1.
+   */
+  @property({ attribute: false })
+  l1Data? = { title: '', url: '', menuItems: [] };
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead-composite.ts:446]: Given we don't want to render L1 when `this.l1Data` is `undefined`, I think we don't want to define the default value:

```suggestion
  l1Data?: MastheadL1;
```

File: packages/web-components/src/components/masthead/masthead-l1-name.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    return `${ddsPrefix}--masthead__l1-name`;
+  }
+
+  render() {
+    return html`
+      <span class="${prefix}--masthead__l1-name-title">
+        <a href="${this.url}">${this.title}</a>
+      </span>
+    `;
+  }
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead-l1-name.ts:48]: Method grouping nit:

```suggestion
  render() {
    return html`
      <span class="${prefix}--masthead__l1-name-title">
        <a href="${this.url}">${this.title}</a>
      </span>
    `;
  }

  static get stableSelector() {
    return `${ddsPrefix}--masthead__l1-name`;
  }
```

File: packages/web-components/src/components/masthead/masthead-l1.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  render() {
+    return html`
+      <slot name="name"></slot>
+      <slot></slot>
+    `
+  }
+
+  static styles = styles; // `styles` here is a `CSSResult` generated by custom WebPack loader
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead-l1.ts:38]: Method grouping and whitespacing nit:

```suggestion
  render() {
    return html`
      <slot name="name"></slot>
      <slot></slot>
    `
  }

  static get stableSelector() {
    return `${ddsPrefix}--masthead__l1`;
  }

  static styles = styles; // `styles` here is a `CSSResult` generated by custom WebPack loader
```

File: packages/web-components/src/components/masthead/masthead.scss [hunk 7]
```diff
@@ -3,19 +3,20 @@
 //
 // This source code is licensed under the Apache-2.0 license found in the
 // LICENSE file in the root directory of this source tree.
-//
+//
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead.scss:6]: Whitespacing nit:

```suggestion
//
```

File: packages/web-components/src/components/masthead/masthead.scss [hunk 8]
```diff
@@ -25,6 +26,7 @@
       overflow-x: visible;
     }
   }
+
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead.scss:29]: Whitespacing nit:

```suggestion
```

File: packages/web-components/src/components/masthead/masthead.scss [hunk 9]
```diff
@@ -3,19 +3,20 @@
...
 @import '@carbon/ibmdotcom-styles/scss/globals/vars';
 @import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead';
 @import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead-leftnav';
 @import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead-search';
+@import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead-l1';
 @import 'carbon-web-components/scss/components/ui-shell/header';
+@import 'carbon-web-components/scss/components/ui-shell/side-nav'; 
 @import 'carbon-web-components/scss/components/ui-shell/side-nav';
 @import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead-megamenu';
 @import '../../globals/scss/box-sizing';
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead.scss:19]: Whitespacing nit:

```suggestion
@import '../../globals/scss/box-sizing';

```

File: packages/web-components/src/components/masthead/masthead.scss [hunk 10]
```diff
@@ -3,19 +3,20 @@
...
 
 @import 'carbon-components/scss/components/ui-shell/header';
 @import 'carbon-components/scss/components/ui-shell/side-nav';
 @import '@carbon/ibmdotcom-styles/scss/globals/vars';
 @import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead';
 @import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead-leftnav';
 @import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead-search';
+@import '@carbon/ibmdotcom-styles/scss/components/masthead/masthead-l1';
 @import 'carbon-web-components/scss/components/ui-shell/header';
+@import 'carbon-web-components/scss/components/ui-shell/side-nav';
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead.scss:16]: Seems to be duplicate:

```suggestion
```

File: packages/web-components/src/components/masthead/masthead.scss [hunk 11]
```diff
@@ -318,4 +320,4 @@
 :host(#{$dds-prefix}-masthead-search-composite),
 :host(#{$dds-prefix}-masthead-search-container) {
   display: contents;
-}
+}
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead.scss:323]: Whitespacing nit:

```suggestion
}

```

File: packages/styles/scss/components/masthead/_masthead-l1.scss [hunk 12]
```diff
@@ -53,6 +55,10 @@ $search-transition-timing: 95ms;
     }
   }
 
+  :host(#{$dds-prefix}-masthead-l1-name) span::slotted(*) {
+    color: pink;
+  }
+
```
Comment (asudoh) [inline packages/styles/scss/components/masthead/_masthead-l1.scss:61]: Seems a debugging code:

```suggestion
```

File: packages/styles/scss/components/masthead/_masthead-l1.scss [hunk 13]
```diff
@@ -38,11 +40,11 @@ $search-transition-timing: 95ms;
     }
   }
 
-  .#{$prefix}--masthead__l1-name {
+  .#{$prefix}--masthead__l1-name, :host(#{$dds-prefix}-masthead-l1-name){
     display: flex;
     height: 100%;
     background-color: $ui-02;
-    color: $text-04;
+    // color: $text-04;
```
Comment (asudoh) [inline packages/styles/scss/components/masthead/_masthead-l1.scss:47]: I guess you meant to restore this CSS rule?

```suggestion
    color: $text-04;
```

**Review by asudoh — COMMENTED** [review]
Getting there @RaphaelAmadeu!

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 14]
```diff
@@ -68,6 +72,85 @@ enum NAV_ITEMS_RENDER_TARGET {
  */
 @customElement(`${ddsPrefix}-masthead-composite`)
 class DDSMastheadComposite extends LitElement {
+  private _renderL1Items({ target }: { target: NAV_ITEMS_RENDER_TARGET }) {
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead-composite.ts:75]: Could you please add JSDoc comment here?

File: packages/web-components/src/components/masthead/masthead-composite.ts [hunk 15]
```diff
@@ -68,6 +72,85 @@ enum NAV_ITEMS_RENDER_TARGET {
...
+                    data-autoid="${ddsPrefix}--masthead__l1-sidenav--nav-${i}"
+                  ></dds-left-nav-item>
+                `
+          );
+    }
+
+    return undefined;
+  }
+
+  private _renderL1() {
```
Comment (asudoh) [inline packages/web-components/src/components/masthead/masthead-composite.ts:139]: Could you please add JSDoc comment here?

**Review by asudoh — APPROVED** [review]
🎉 - Also, could you look at the failed CI to see what files needs to be re-formatted by Prettier? Thanks @RaphaelAmadeu!

File: packages/web-components/src/components/masthead/masthead-composite.ts (hunk 14 above)
Comment (asudoh) [reply packages/web-components/src/components/masthead/masthead-composite.ts:75]: (This is for striving for easier future maintenance with well-documented codebase)

File: packages/web-components/src/components/masthead/masthead-composite.ts (hunk 15 above)
Comment (asudoh) [reply packages/web-components/src/components/masthead/masthead-composite.ts:143]: (This is for striving for easier future maintenance with well-documented codebase)

**Review by kennylam — APPROVED** [review]
Looks good thanks @RaphaelAmadeu!

**Pull request merged**